Update WABS azure-storage blob API to 0.34.2

nkiraly opened the pull request:
There have been several breaking changes to azure-storage APIs across the Azure SDK 2.0.0 release candidates and the component versions they tracked.

Update WAL-E minimum Azure SDK support to version 2.0.0 to ensure users in Azure are able to leverage WABS.
* migrate uri_put_file from put_blob / put_block / put_block_list to create_blob_from_bytes which has built in segmenting and parallelization
* migrate uri_get_file from get_blob with chunking to get_blob_to_stream which has built in segmenting
* update extras to azure 2.0.0
* update azure-storage minimum version to 0.34.2

    $ tox -e py35
    123 passed, 129 skipped in 20.06 seconds

More Info:
https://github.com/Azure/azure-storage-python/blob/master/BreakingChanges.md#version-0300
https://azure-sdk-for-python.readthedocs.io/en/v0.11.1/ref/azure.storage.blobservice.html
https://docs.microsoft.com/en-us/python/api/azure.storage.blob.blockblobservice.blockblobservice?view=azure-python#azure_storage_blob_blockblobservice_BlockBlobService_create_blob_from_bytes
https://docs.microsoft.com/en-us/python/api/azure.storage.blob.baseblobservice.baseblobservice?view=azure-python#azure_storage_blob_baseblobservice_BaseBlobService_list_blobs
https://github.com/Azure/azure-storage-python/blob/master/samples/blob/block_blob_usage.py

titanous commented:
This also has a side effect of fixing a bug where any backups past the first 14 were inaccessible due to the list page size. The new default `list_blobs` size appears to be 5000, and it might make sense to specify that explicitly.

titanous commented:
This branch appears to leak file descriptors during a `backup-fetch`, eventually resulting in EMFILE errors while downloading new segments.

This is the list of fds with a pool size of 10 after about 20 segments: https://gist.github.com/titanous/3808052e8c7e0cefa4009adebe4d6a4c

The number of open file descriptors grew with each new segment that started downloading.

bearrito commented:
Does a specific test show those leaks?

titanous commented:
It happens consistently when downloading a large base backup. You could make a large dataset with pgtune to test.

nkiraly commented:
I am looking at making further changes to make sure this works reliably.

nkiraly commented:
Using a package built with 5ae3deb I was able to backup-push a PostgreSQL database server with 20GB of databases to Azure blob storage, and restore that backup-fetch that basebackup to a new server.

I advocate for this pull request to be reviewed for acceptance. Any feedback is appreciated.

bearrito commented:
@titanous Is there anything else required for this?

deverant commented:
Hey! I assume you have tested this in production so the change is good for Azure and works in production? Really like how the never libraries allows for some code deletion. Two things come to mind right now:

1. Can you rebase the branch and squash some of these commits?
2. Have you been able to test this with another storage? Patching threads makes me slightly nervous since it can affect the other libraries used for other storages. I can see if I have time to test this branch out against GCS at least but we should also test it against AWS for sure.

nkiraly commented:
1. Yes I have been using this in production for some time. Base backups, WAL archiving, and restoration working as expected.

2. I have not tested these changes with other storage systems.

3. I have rebased the branch and squashed the SDK update iterations. Tests still passing locally with 5e0a415, and Travis CI agrees.

deverant commented:
Verified that integration tests for GCS work even after the threads are patched.

deverant commented:
@fdr Merging this probably means we need a version bump. Any thoughts on this? We should probably release a new version anyway now that the python 3.6 errors are fixed.

vijayrajah commented:
I have used @nkiraly with python 3.6 on Centos7. It is working without any issues. (yet!). I still need to test the restore part. 

Can you please merge this and create a new version.

fdr commented:
Let's go ahead and release.

On Wed, Jul 11, 2018 at 4:09 AM vijayrajah <notifications@github.com> wrote:

> I have used @nkiraly <https://github.com/nkiraly> with python 3.6 on
> Centos7. It is working without any issues. (yet!). I still need to test the
> restore part.
>
> Can you please merge this and create a new version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wal-e/wal-e/pull/378#issuecomment-404132225>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAcF-hdmzwK4Fv3YMJgEHcWMQIdgB_7ks5uFdzbgaJpZM4R9dj_>
> .
>
